python/ray/serve/_private/http_proxy.py

```diff
@@ -650,6 +671,79 @@ async def proxy_asgi_receive(
             if msg["type"] == "websocket.disconnect":
                 return msg["code"]
 
+    async def _assign_request_with_timeout(
+        self,
+        handle: RayServeHandle,
+        scope: Scope,
+        disconnected_task: asyncio.Task,
+        timeout_s: Optional[float] = None,
+    ) -> Optional[StreamingObjectRefGenerator]:
+        """Attempt to send a request on the handle within the timeout.
+
+        If `timeout_s` is exceeded while trying to assign a replica, `TimeoutError`
+        will be raised.
+
+        `disconnected_task` is expected to be done if the client disconnects; in this
+        case, we will abort assigning a replica and return `None`.
+        """
+        assignment_task = handle.remote(pickle.dumps(scope), self.self_actor_handle)
+        done, _ = await asyncio.wait(
+            [assignment_task, disconnected_task],
+            return_when=FIRST_COMPLETED,
+            timeout=timeout_s,
+        )
+        if disconnected_task in done:
+            assignment_task.cancel()
+            return None
+        elif assignment_task in done:
+            return assignment_task.result()
+        else:
+            raise TimeoutError()
```

Comment:
I think you have to kill "pending" tasks (which is "_" now). Looks like a leak to me.